xarray/core/dataarray.py

```diff
@@ -4490,15 +4490,17 @@ def query(
         Dimensions without coordinates: x
         """
 
-        ds = self._to_dataset_whole(shallow_copy=True)
+        name = _THIS_ARRAY if self.name is None else self.name
+        ds = self._to_dataset_whole(name=name, shallow_copy=True)
```

Comment:
is there a reason why we don't just use `_to_temp_dataset` / `_from_temp_dataset` regardless of the name?

Reply:
`_to_dataset_whole` is just the same thing as `_to_temp_dataset` except it accepts the `shallow_copy` argument, which is was already being used here.

Using `_to_dataset_whole` with `name=None` will fail (so I can't just pass `self.name` because that will be `None` for an unnamed dataarray). Using `_to_dataset_whole` without passing a name fails in the same way.

Then trying to use `_from_temp_dataset` fails with

```python
    def _from_temp_dataset(
        self, dataset: Dataset, name: Union[Hashable, None, Default] = _default
    ) -> "DataArray":
>       variable = dataset._variables.pop(_THIS_ARRAY)
E       KeyError: <this-array>
```
so I think it will fail for named dataarrays?

Reply:
I don't think we need the `shallow_copy=True` argument as this is a temporary dataset (*and* it's the usual pattern for delegating to the `Dataset` implementation). No need to use the `name` parameter for `_from_temp_dataset`, either, as it's being restored from `self`.

Reply:
>  I'd suggest to use the same pattern as in `sel`

That doesn't work for `query` because `query` needs the name of the dataarray to be a variable on the temporary dataset, otherwise it can't evaluate queries involving the dataarray name (such as `x="a > 5"` where `da.name = 'a'`). The tests will fail with this error if you try that:

```python
>               raise UndefinedVariableError(key, is_local) from err
E               pandas.core.computation.ops.UndefinedVariableError: name 'a' is not defined
```

Reply:
Basically in order for query to work you have to provide a name to construct the temporary dataset with, which rules out using `_to_temp_dataset`.

Reply:
I see. `pandas` skipped this issue completely: `Series` objects don't define `.query`. In this [feature request](https://github.com/pandas-dev/pandas/issues/22347) someone proposed using `self` as a reference to `DataArray`, which could be easily done by passing `"self"` to `_to_dataset_whole` (without the shallow copy) and then renaming the result. I wonder if that's a better than relying on the name of the `DataArray`, which might be undefined?

Reply:
Interesting.

> I wonder if that's a better than relying on the name of the DataArray, which might be undefined?

I don't think we should use `'self'` always, because then in the case of named dataarrays we would have deviated from the behaviour of `dataset.query()`: always querying with self is mutually exclusive with querying with the name of the variable in a named datarray.

Using `'self'` instead of `'<this-array>'` is a cool idea for unnamed dataarrays though. But it doesn't actually solve the `_to/_from` problem you originally asked about, it just adds the ability for the query to depend on the values of the unnamed datarray.

Reply:
I added the ability to reference as `'self'` but it doesn't make the underlying implementation any cleaner :sweat_smile:

Reply:
indeed. However, if I had to choose between `"self"` for unnamed and the name for named `DataArray` objects or always `"self"` I'd choose the latter because it's easier to use (and the implementation would be easier)? Not sure if we should try to support both (i.e. allow referencing by name and using `"self"`)?

Reply:
I see what you mean. What I just wrote supports both individually: You use
the name when it's named, and self when it isn't. But if it's named you
can't use self. I suppose you could support using self even when it's named
by examining the query and replacing parts of it before giving it to
ds.query?

Reply:
So I added support for using `'self'` even when the dataarray has a name, by using `.replace()`. It works, but makes the implementation more complicated again.

Reply:
One way to slightly simplify this would be to always rename all dataarray names to self before querying.

Reply:
This will break if there's a non-dimensional coordinate variable named `self` won't it? I'm not sure there's  good solution though, perhaps `...` :D